src/tcp.h

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2020 Paul-Louis Ageneau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#ifndef JUICE_TCP_H
+#define JUICE_TCP_H
+
+#include "addr.h"
+#include "socket.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+socket_t tcp_create_socket(const addr_record_t *dst) {
+	socket_t tcp_socket = socket(AF_INET, SOCK_STREAM, 0);
+	if (tcp_socket == -1) {
+		return INVALID_SOCKET;
+	}
+
+	ctl_t nbio = 1;
+	if (ioctlsocket(tcp_socket, FIONBIO, &nbio)) {
+		closesocket(tcp_socket);
+		return SECONNREFUSED;
+	}
+
+	int ret = connect(tcp_socket, (const struct sockaddr *)&dst->addr, dst->len);
+	if (ret != 0 && sockerrno != SEINPROGRESS && sockerrno != SEWOULDBLOCK) {
+		closesocket(tcp_socket);
+		return SECONNREFUSED;
+	}
+
+	return tcp_socket;
+}
+
+int tcp_ice_write(socket_t sock, const char *data, size_t size) {
+	if (size >= USHRT_MAX) {
+		return SEMSGSIZE;
+	}
+
+	uint16_t header = htons((uint16_t) size);
+	int ret = (int)send(sock, (const char *) &header, sizeof(uint16_t), 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return send(sock, data, (int)size, 0);
+}
+
+int tcp_ice_read(socket_t sock, char *buffer, size_t size, uint16_t *ice_tcp_len) {
+	int ret = 0;
+
+	if (*ice_tcp_len == 0) {
+		if ((ret = recv(sock, (char *)ice_tcp_len, sizeof(uint16_t), 0)) != 2) {
+			goto __exit;
+		}
+
+		*ice_tcp_len = ntohs(*ice_tcp_len);
+		if (size < *ice_tcp_len) {
+			return -1;
+		}
+	}
+
+	if ((ret = recv(sock, buffer, *ice_tcp_len, 0)) == *ice_tcp_len) {
+		*ice_tcp_len = 0;
+	}
```

Comment:
The logic here seems incorrect as it does not handle short reads properly. What if `recv()` returns less than `*ice_tcp_len`? It looks like in that case `*ice_tcp_len` is not decremented, so the next call won't know part of the message has been read already and framing will be broken, and `tcp_ice_read` returns a truncated message, so processing will fail. Am I missing something here?

Reply:
Nope you were exactly right! I need a peek. Doing partials reads would have done extra copies, so this works out for the best.